waiter/src/waiter/discovery.clj

```diff
@@ -89,14 +104,39 @@
     (zipmap (map #(str (.getId %)) filtered-instances)
             (map #(get-instance-url % protocol endpoint) filtered-instances))))
 
+(defn- router-details
+  "Returns a map representing the details of the given router."
+  [instance]
+  (let [custom-details (pc/map-keys keyword (.getPayload instance))
+        details {:address (.getAddress instance)
+                 :custom-details custom-details
+                 :id (.getId instance)
+                 :name (.getName instance)
+                 :port (.getPort instance)}]
+    details))
+
+(defn router-id->details
+  "For all routers that pass the exclude-set, returns a mapping from router-id to its details."
+  [discovery & {:keys [exclude-set] :or {exclude-set #{}}}]
+  (let [filtered-instances (routers discovery exclude-set)]
+    (zipmap (map #(str (.getId %)) filtered-instances)
+            (map router-details filtered-instances))))
+
 (defn cluster-size
   "Returns the number of routers particpating in the ZooKeeper cluster."
   [discovery]
   (count (routers discovery #{})))
 
+(defn versioned-discovery-path
+  "Returns a versioned discovery path. See the note in ->service-instance about payload types
+   and discovery paths."
+  [discovery-path]
+  (str discovery-path "-v2"))
+
 (defn register
-  [router-id curator service-name discovery-path {:keys [host port]}]
-  (let [instance (->service-instance router-id service-name {:host host :port port})
+  [router-id curator service-name discovery-path-base {:keys [host port router-fqdn router-ssl-port]}]
+  (let [discovery-path (versioned-discovery-path discovery-path-base)
+        instance (->service-instance router-id service-name {:host host :port port :router-fqdn router-fqdn :router-ssl-port router-ssl-port})
```

Comment:
For backward compatibility should the router register to both the old and `v2` paths?

Reply:
No, I don't think so. The reason "-v2" is introduced is that the existing service instances and the new one are incompatible. If this new service, using a HashMap payload, also registered with the old path, then on startup of a new router, the existing routers would discover it and fail trying to deserialize its definition.

Reply:
We discussed offline nd decided to avoid registering to both locations as the peers not discovering each other would be expected only in a temporary state.